File: backend/app/schemas/pricing.py
"""Pydantic schemas for the pricing calculation API endpoints."""
from pydantic import BaseModel, Field
from typing import Any, Dict, List, Optional


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------

class AddonsInput(BaseModel):
    rush: bool = False
    double_rush: bool = False
    # Dance voiceover amount: 25, 75, or 100 (confirm with Megan on $100 tier)
    dance_voiceover_amount: Optional[int] = None
    # Cheer voiceover amount: 20 or 40
    cheer_voiceover_amount: Optional[int] = None
    eight_count_sheets: bool = False
    extra_songs: int = 0
    extra_song_added_time: int = 0


class PricingCalculateRequest(BaseModel):
    form_type: str
    cheer_form_subtype: Optional[str] = None
    dance_form_subtype: Optional[str] = None
    package_name: str
    music_affiliate: Optional[str] = None
    addons: Optional[AddonsInput] = None


class CompletePricingRequest(BaseModel):
    """Body for POST /orders/{id}/complete-pricing"""
    # If not provided, calculated from classification
    cheer_form_subtype: Optional[str] = None
    dance_form_subtype: Optional[str] = None
    music_affiliate: Optional[str] = None
    addons: Optional[AddonsInput] = None
    # Optional: Megan's price override (if she sets a different customer-facing price)
    final_customer_price_override: Optional[float] = None


class FinalizePayrollRequest(BaseModel):
    """Body for POST /orders/{id}/finalize-payroll"""
    producer_initials: str
    # Optional: override the auto-calculated final_customer_price for SLT portion
    final_customer_price: Optional[float] = None
    # Optional: override the auto-calculated rate (e.g. Megan selects old vs new for Casey)
    overridden_rate: Optional[float] = None


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------

class AddOnLineItemResponse(BaseModel):
    addon_id: str
    label: str
    customer_amount: float
    payroll_amount: float
    quantity: int
    note: Optional[str] = None


class PricingBreakdownResponse(BaseModel):
    form_type: str
    canonical_subtype_id: Optional[str]
    package_id: Optional[str]
    package_name: Optional[str]
    pricing_rule_id: Optional[str]

    compliance_status: str
    compliance_reason: str
    canonical_affiliate: Optional[str]

    base_customer_price: Optional[float]
    base_payroll_price: Optional[float]

    addons: List[AddOnLineItemResponse] = []

    system_calculated_customer_price: Optional[float]
    payroll_base_price: Optional[float]

    needs_manual_pricing: bool
    needs_manual_review: bool
    summary_line: str


class CompensationResultResponse(BaseModel):
    status: str
    producer_payout: Optional[float]
    rate_used: Optional[float]
    rate_source: str
    old_pricing_payout: Optional[float] = None
    new_pricing_payout: Optional[float] = None
    old_new_trigger_unconfirmed: bool = False
    manual_input_fields: List[Dict[str, Any]] = []
    slt_portion: Optional[float]
    final_customer_price: Optional[float]
    flag: str


class OrderPricingResponse(BaseModel):
    """Returned by complete-pricing and finalize-payroll endpoints."""
    order_id: str
    # Pricing fields
    system_calculated_customer_price: Optional[float]
    final_customer_price: Optional[float]
    final_customer_price_overridden: bool
    price_compliance: Optional[str]
    pricing_breakdown: Optional[Dict[str, Any]]
    # Payroll fields (only set after finalize-payroll)
    rate_used: Optional[float] = None
    rate_source: Optional[str] = None
    producer_payout: Optional[float] = None
    slt_portion: Optional[float] = None
    payroll_finalized: bool = False
    payroll_breakdown: Optional[Dict[str, Any]] = None

File: src/lib/api/pricing.ts
import { apiClient } from "./client";

export interface AddOnLineItem {
  addon_id: string;
  label: string;
  customer_amount: number;
  payroll_amount: number;
  quantity: number;
  note: string | null;
}

export interface PricingBreakdown {
  form_type: string;
  canonical_subtype_id: string | null;
  package_id: string | null;
  package_name: string | null;
  pricing_rule_id: string | null;
  compliance_status: "compliant" | "non-compliant" | "needs_manual_review";
  compliance_reason: string;
  canonical_affiliate: string | null;
  base_customer_price: number | null;
  base_payroll_price: number | null;
  addons: AddOnLineItem[];
  system_calculated_customer_price: number | null;
  payroll_base_price: number | null;
  needs_manual_pricing: boolean;
  needs_manual_review: boolean;
  summary_line: string;
}

export interface CompensationResult {
  status: "computed" | "needs_manual_review" | "hourly_manual" | "not_paid_for_mixing";
  producer_payout: number | null;
  rate_used: number | null;
  rate_source: string;
  old_pricing_payout?: number | null;
  new_pricing_payout?: number | null;
  old_new_trigger_unconfirmed?: boolean;
  manual_input_fields?: Array<{ label: string; rate_or_null: number | null }>;
  slt_portion: number | null;
  final_customer_price: number | null;
  flag: string;
}

export interface OrderPricingResponse {
  order_id: string;
  system_calculated_customer_price: number | null;
  final_customer_price: number | null;
  final_customer_price_overridden: boolean;
  price_compliance: string | null;
  pricing_breakdown: PricingBreakdown | null;
  rate_used?: number | null;
  rate_source?: string | null;
  producer_payout?: number | null;
  slt_portion?: number | null;
  payroll_finalized?: boolean;
  payroll_breakdown?: CompensationResult | null;
}

export interface AddonsInput {
  rush?: boolean;
  double_rush?: boolean;
  dance_voiceover_amount?: number | null;
  cheer_voiceover_amount?: number | null;
  eight_count_sheets?: boolean;
  extra_songs?: number;
  extra_song_added_time?: number;
}

export interface PricingCalculateRequest {
  form_type: string;
  cheer_form_subtype?: string | null;
  dance_form_subtype?: string | null;
  package_name: string;
  music_affiliate?: string | null;
  addons?: AddonsInput;
}

export interface CompletePricingRequest {
  cheer_form_subtype?: string | null;
  dance_form_subtype?: string | null;
  music_affiliate?: string | null;
  addons?: AddonsInput;
  final_customer_price_override?: number | null;
}

export interface FinalizePayrollRequest {
  producer_initials: string;
  final_customer_price?: number | null;
  overridden_rate?: number | null;
}

export async function calculatePricingApi(
  req: PricingCalculateRequest
): Promise<PricingBreakdown> {
  return apiClient.post<PricingBreakdown>("/api/pricing/calculate", req);
}

export async function completePricingApi(
  orderId: string,
  req: CompletePricingRequest
): Promise<OrderPricingResponse> {
  return apiClient.post<OrderPricingResponse>(
    `/api/orders/${orderId}/complete-pricing`,
    req
  );
}

export async function finalizePayrollApi(
  orderId: string,
  req: FinalizePayrollRequest
): Promise<OrderPricingResponse> {
  return apiClient.post<OrderPricingResponse>(
    `/api/orders/${orderId}/finalize-payroll`,
    req
  );
}
